reco_utils/recommender/fbt/fbt.py

```diff
@@ -0,0 +1,311 @@
+        )
+        # similarity score is defined by how many distinct
+        # users interacted with the same pair of items
+        sim_df = (
+            cooccurrence_df
+            .groupby([self.col_item,
+                     f'{self.col_item}_paired'])[self.col_user]
+            .nunique()
+            .reset_index(drop=False)
+            .rename(columns={self.col_user: self.col_score})
+        )
+
+        # Retaining item-pairs where item isn't paired with itself
+        fbt_df = (
+            sim_df
+            .loc[sim_df
+                 [self.col_item] !=
+                 sim_df[f'{self.col_item}_paired']]
+        )
+
+        self._model_df = fbt_df
+
+        # Item frequencies can be obtained by looking at the
+        # number of distinct users who interacted with a specific
+        # item: can be extracted from items paired with itself
+        self.item_frequencies = (
+            sim_df
+            .loc[sim_df
+                 [self.col_item] ==
+                 sim_df[f'{self.col_item}_paired']]
+            .drop(columns=[f'{self.col_item}_paired'])
+        )
+
+        self._is_fit = True
+        logger.info("Done training")
+
+    def predict(self, test):
+        """Generate new recommendations using a trained FBT model.
+
+        Parameters
+        ----------
+        test : DataFrame
+            DataFrame of users who need to get recommendations
+        remove_seen (bool): flag to remove items seen in training
+        from recommendation
+
+        Returns
+        -------
+        R : DataFrame
+            DataFrame with up to `num_recos` rows per user in X.
+        """
+        if not self._is_fit:
+            raise ValueError(("fit() must be called before predict()!"))
+
+        self._check_dataframe(test)
+        logger.info("Calculating recommendation scores")
+
+        # To get recommendations, merge test dataset of users with the
+        # item similarity on col_item to get all matches
+        all_test_item_matches_df = test.merge(
+            self._model_df,
+            on=self.col_item,
+            how='left'
+        )
+        # Give user may interact with multiple items A, B, C, a
+        # item D may be recommended as a popular paired choice for
+        # each of A, B, C. Will average such scores
+        all_recommendations_df = (
+            all_test_item_matches_df
+            .groupby([self.col_user,
+                     f'{self.col_item}_paired'])[self.col_score]
+            .mean()
+            .reset_index(drop=False)
+            .rename(columns={f'{self.col_item}_paired': self.col_item})
+        )
+        all_recommendations_df[self.col_item] = (
+            all_recommendations_df[self.col_item].astype('int64')
+        )
+
+        return all_recommendations_df
+
+    def recommend_k_items(self,
+                          test,
+                          remove_seen=False,
+                          train=None,
+                          top_k=None):
```

Comment:
in other functions like this we put `top_k=10` or the default constant: https://github.com/microsoft/recommenders/blob/main/reco_utils/recommender/sar/sar_singlenode.py#L450

Reply:
sounds good...topk default set to 10.